include/zephyr/bluetooth/conn.h

```diff
@@ -1173,8 +1173,11 @@ struct bt_conn_cb {
 	 * @ref bt_conn_unref which is used by the BT stack. Making
 	 * Bluetooth API calls in this context is error-prone and strongly
 	 * discouraged.
+	 *
+	 * @param conn Connection object that was returned. Can ONLY be used
+	 *             for comparison purposes, not API calls.
 	 */
-	void (*recycled)(void);
+	void (*recycled)(const void *conn);
```

Comment:
Would it be possible to get some sample code for how exactly you see this being used? It sounds like a bug to me if the application holds a pointer to a connection object but doesn't actually own a valid reference to it (which contributes to the reference count). If it does own a reference then the connection can't be recycled, and if it doesn't have such a pointer then what is there to compare against?

Reply:
```
static const struct bt_le_ext_adv_cb adv_cb = {
	.connected = adv_set_connected,
};
static struct bt_conn *legacy_conn;

BT_CONN_CB_DEFINE(conn_cb) = {
	.recycled = conn_recycled_cb,
};

static void adv_set_connected(struct bt_le_ext_adv *adv, struct bt_le_ext_adv_connected_info *info)
{
	legacy_conn = info->conn;
}

static void conn_recycled_cb(const void *conn)
{
	if (conn != legacy_conn) {
		return;
	}
	legacy_conn = NULL;
	// Schedule work to restart advertising
	k_work_schedule(&restart_advertising, K_NO_WAIT);
}
```
As mentioned in the description I am looking to replicate `!BT_LE_ADV_OPT_ONE_TIME` for extended advertising.
The existing recycled callback is insufficient as there are other advertising sets that exist in the application.

Reply:
I also don't understand your use case. Could you describe it a bit more, what's the logic for starting your advertiser and what are the other advertising sets used for?

Reply:
So iOS devices cannot connect to devices running the Nordic Softdevice over extended advertising per [this issue (DRGN-22053)](https://devzone.nordicsemi.com/f/nordic-q-a/113141/softdevice-extended-advertising-ios-connection).

I use extended advertising for data broadcast, but because of the above, I require a second advertising set that operates only on the legacy channels only to enable iOS. I therefore want to start the legacy set at boot, and restart it whenever a connection that resulted from the set terminates. Simply starting the legacy set every time `recycled` is called is incorrect, because `recycled` is called every time a connectable advertising set terminates.

Reply:
Having no way to identify the connection in `recycled` assumes that all recycling events are interchangeable, which is not correct when you want a library that is responsible for a single advertising set only.

Reply:
In the above code, shouldn't

```c
legacy_conn = info->conn;
```
have been
```c
legacy_conn = bt_conn_ref(info->conn);
```

As @jhedberg pointed out? 
Storing and comparing a `bt_conn` without taking a reference is a problem waiting to happen.

Reply:
I agree with @JordanYates, I see no way in our current API to do what he describes exactly. I understand the goal is a "nice neighbor" library that does not allocate a new connection before it is sure the previous connection object it allocated is gone, even if it is some other part of the system that is keeping that object alive.

Reply:
> In the above code, shouldn't
> 
> ```c
> legacy_conn = info->conn;
> ```
> 
> have been
> 
> ```c
> legacy_conn = bt_conn_ref(info->conn);
> ```
> 
> As @jhedberg pointed out? Storing and comparing a `bt_conn` without taking a reference is a problem waiting to happen.

No, because if a reference to the connection is held the `recycled` callback will never be run